OPAM executes `dune build -p ''`  when a package is rebuilt but not in the current repo scope.l

We have encountered a bit tricky bug of OPAM.

1. Install a package A in the default repo.
2. Remove the default and add a custom repo R without A.
3. Installation of a package in R which triggers recompilation of A
4. A is not in the current scope, but OPAM has its source locally.
5. Reinstallation of A fails saying:

```
[ERROR] The compilation of res failed at "/home/MINE/.opam/opam-init/hooks/sandbox.sh build dune build -p  -j 4".
```

Note that `dune`'s `-p` option is given an empty string: `dune build -p '' -j 4` is the executed command.  The crash of `dune` command is a bug reported at: https://github.com/ocaml/dune/issues/3530

OPAM has its own bug: it should reinstall package A with a proper package name to `dune`.  Otherwise it should properly fail complaining it cannot get the package name to build.

Currently, we have worked around the issue by tweaking versions of packages to be installed so that A is never recompiled at step 3.

--- a/master_changes.md
+++ b/master_changes.md
@@ -20,6 +20,7 @@ note.
 
 ## Switch
   * Fix Not_found with `opam switch create . --deps` [#4151 @AltGr]
+  * Package Var: resolve self `name` variable for orphan packages [#4228 @rjbou - fix #4224]
 
 ## Pin
   * Don't keep unpinned package version if it exists in repo [#4073 @rjbou - fix #3630]

--- a/src/state/opamPackageVar.ml
+++ b/src/state/opamPackageVar.ml
@@ -260,8 +260,11 @@ let resolve st ?opam:opam_arg ?(local=OpamVariable.Map.empty) v =
       Some (bool false)
     | "pinned", _ ->
       Some (bool (OpamPackage.has_name st.pinned name))
-    | "name", _ ->
-      if OpamPackage.has_name st.packages name
+    | "name", opam ->
+      (* On reinstall, orphan packages are not present in the state, and we
+         need to resolve their internal name variable *)
+      if OpamStd.Option.map OpamFile.OPAM.name opam = Some name
+      || OpamPackage.has_name st.packages name
       then Some (string (OpamPackage.Name.to_string name))
       else None
     | _, None -> None

